File: server/src/controllers/judge0.controllers.js
import { compileWithJudge0 } from '../services/judge0.service.js';
import { JUDGE0_BASE_URL, judge0Headers } from "../config/judge0.config.js";
import axios from 'axios';

export const compileCode = async (req, res, next) => {
  try {
    const output = await compileWithJudge0(req.body);
    res.json({ success: true, output });
  } catch (err) {
    next(err);
  }
};

export const checkJudge0Status = async (req, res, next) => {
  try {
    const response = await axios.get(`${JUDGE0_BASE_URL}/about`, { headers: judge0Headers });
    res.json({ success: true, data: response.data });
  } catch (err) {
    next(err);
  }
};

File: server/src/controllers/judge0.controllers.go
package controllers

import (
	"net/http"

	"zstudio/server/src/httpx"
	"zstudio/server/src/schema"
	"zstudio/server/src/services"
)

type Judge0Controller struct {
	Service *services.Judge0Service
}

func NewJudge0Controller(service *services.Judge0Service) *Judge0Controller {
	return &Judge0Controller{Service: service}
}

func (c *Judge0Controller) CompileCode(w http.ResponseWriter, r *http.Request) {
	body, err := httpx.ReadJSONBody(r)
	if err != nil {
		httpx.WriteError(w, httpx.AppError{Status: http.StatusBadRequest, Message: err.Error()})
		return
	}

	compilePayload, validationError := schema.ValidateJudge0Payload(body)
	if validationError != "" {
		httpx.WriteJSON(w, http.StatusBadRequest, map[string]any{
			"success": false,
			"error":   validationError,
		})
		return
	}

	output, err := c.Service.CompileWithJudge0(r.Context(), compilePayload)
	if err != nil {
		httpx.WriteError(w, httpx.AppError{Status: http.StatusInternalServerError, Message: err.Error()})
		return
	}

	httpx.WriteJSON(w, http.StatusOK, map[string]any{
		"success": true,
		"output":  output,
	})
}

func (c *Judge0Controller) CheckJudge0Status(w http.ResponseWriter, r *http.Request) {
	data, err := c.Service.CheckSystemInfo(r.Context())
	if err != nil {
		httpx.WriteError(w, httpx.AppError{Status: http.StatusInternalServerError, Message: err.Error()})
		return
	}

	httpx.WriteJSON(w, http.StatusOK, map[string]any{
		"success": true,
		"data":    data,
	})
}
